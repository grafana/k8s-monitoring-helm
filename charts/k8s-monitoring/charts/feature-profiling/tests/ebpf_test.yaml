# yamllint disable rule:document-start rule:line-length rule:trailing-spaces
suite: Test eBPF profiling
templates:
  - configmap.yaml
tests:
  - it: should build the eBPF profiling configuration
    set:
      deployAsConfigMap: true
      ebpf:
        enabled: true
      java:
        enabled: false
      pprof:
        enabled: false
    asserts:
      - isKind:
          of: ConfigMap
      - matchSnapshot:
          path: data["module.alloy"]

  - it: should be able to filter by namespace and extra discovery rules
    set:
      deployAsConfigMap: true
      ebpf:
        enabled: true
        namespaces: [default]
        extraDiscoveryRules: |-
          rule {
            source_labels = ["__meta_kubernetes_pod_label_app_kubernetes_io_name"]
            regex = "secret-program"
            action = "drop"
          }
      java:
        enabled: false
      pprof:
        enabled: false
    asserts:
      - isKind:
          of: ConfigMap
      - matchSnapshot:
          path: data["module.alloy"]

  - it: should be able to filter by label and annotation
    set:
      deployAsConfigMap: true
      ebpf:
        enabled: true
        labelSelectors:
          app.kubernetes.io/name: secret-program
          colors: [blue, green]
        annotationSelectors:
          k8s.grafana.com/profile: "true"
      java:
        enabled: false
      pprof:
        enabled: false
    asserts:
      - isKind:
          of: ConfigMap
      - matchSnapshot:
          path: data["module.alloy"]
