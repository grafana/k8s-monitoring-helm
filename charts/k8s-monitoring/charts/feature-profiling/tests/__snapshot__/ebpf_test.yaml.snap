should be able to filter by label and annotation:
  1: |
    |-
      declare "profiling" {
        argument "profiles_destinations" {
          comment = "Must be a list of profile destinations where collected profiles should be forwarded to"
        }
        // Profiles: eBPF
        discovery.kubernetes "ebpf_pods" {
          role = "pod"
          selectors {
            role = "pod"
            label = "app.kubernetes.io/name=secret-program,colors in (blue,green)"
            field = "spec.nodeName=" + sys.env("HOSTNAME")
          }
        }

        discovery.relabel "ebpf_pods" {
          targets = discovery.kubernetes.ebpf_pods.targets
          rule {
            source_labels = ["__meta_kubernetes_pod_phase"]
            regex = "Succeeded|Failed|Completed"
            action = "drop"
          }
          rule {
            source_labels = ["__meta_kubernetes_namespace"]
            target_label = "namespace"
          }
          rule {
            source_labels = ["__meta_kubernetes_pod_annotation_k8s_grafana_com_profile"]
            regex = "true"
            action = "keep"
          }
          rule {
            source_labels = ["__meta_kubernetes_pod_name"]
            target_label = "pod"
          }
          rule {
            source_labels = ["__meta_kubernetes_pod_node_name"]
            target_label = "node"
          }
          rule {
            source_labels = ["__meta_kubernetes_pod_container_name"]
            target_label = "container"
          }
          // provide arbitrary service_name label, otherwise it will be set to {__meta_kubernetes_namespace}/{__meta_kubernetes_pod_container_name}
          rule {
            source_labels = ["__meta_kubernetes_namespace", "__meta_kubernetes_pod_container_name"]
            separator = "@"
            regex = "(.*)@(.*)"
            replacement = "ebpf/${1}/${2}"
            target_label = "service_name"
          }
        }

        pyroscope.ebpf "ebpf_pods" {
          targets = discovery.relabel.ebpf_pods.output
          demangle = "none"
          forward_to = argument.profiles_destinations.value
        }
      }
should be able to filter by namespace and extra discovery rules:
  1: |
    |-
      declare "profiling" {
        argument "profiles_destinations" {
          comment = "Must be a list of profile destinations where collected profiles should be forwarded to"
        }
        // Profiles: eBPF
        discovery.kubernetes "ebpf_pods" {
          role = "pod"
          selectors {
            role = "pod"
            field = "spec.nodeName=" + sys.env("HOSTNAME")
          }
          namespaces {
            names = ["default"]
          }
        }

        discovery.relabel "ebpf_pods" {
          targets = discovery.kubernetes.ebpf_pods.targets
          rule {
            source_labels = ["__meta_kubernetes_pod_phase"]
            regex = "Succeeded|Failed|Completed"
            action = "drop"
          }
          rule {
            source_labels = ["__meta_kubernetes_namespace"]
            target_label = "namespace"
          }
          rule {
            source_labels = ["__meta_kubernetes_pod_name"]
            target_label = "pod"
          }
          rule {
            source_labels = ["__meta_kubernetes_pod_node_name"]
            target_label = "node"
          }
          rule {
            source_labels = ["__meta_kubernetes_pod_container_name"]
            target_label = "container"
          }
          // provide arbitrary service_name label, otherwise it will be set to {__meta_kubernetes_namespace}/{__meta_kubernetes_pod_container_name}
          rule {
            source_labels = ["__meta_kubernetes_namespace", "__meta_kubernetes_pod_container_name"]
            separator = "@"
            regex = "(.*)@(.*)"
            replacement = "ebpf/${1}/${2}"
            target_label = "service_name"
          }
          rule {
            source_labels = ["__meta_kubernetes_pod_label_app_kubernetes_io_name"]
            regex = "secret-program"
            action = "drop"
          }
        }

        pyroscope.ebpf "ebpf_pods" {
          targets = discovery.relabel.ebpf_pods.output
          demangle = "none"
          forward_to = argument.profiles_destinations.value
        }
      }
should build the eBPF profiling configuration:
  1: |
    |-
      declare "profiling" {
        argument "profiles_destinations" {
          comment = "Must be a list of profile destinations where collected profiles should be forwarded to"
        }
        // Profiles: eBPF
        discovery.kubernetes "ebpf_pods" {
          role = "pod"
          selectors {
            role = "pod"
            field = "spec.nodeName=" + sys.env("HOSTNAME")
          }
        }

        discovery.relabel "ebpf_pods" {
          targets = discovery.kubernetes.ebpf_pods.targets
          rule {
            source_labels = ["__meta_kubernetes_pod_phase"]
            regex = "Succeeded|Failed|Completed"
            action = "drop"
          }
          rule {
            source_labels = ["__meta_kubernetes_namespace"]
            target_label = "namespace"
          }
          rule {
            source_labels = ["__meta_kubernetes_pod_name"]
            target_label = "pod"
          }
          rule {
            source_labels = ["__meta_kubernetes_pod_node_name"]
            target_label = "node"
          }
          rule {
            source_labels = ["__meta_kubernetes_pod_container_name"]
            target_label = "container"
          }
          // provide arbitrary service_name label, otherwise it will be set to {__meta_kubernetes_namespace}/{__meta_kubernetes_pod_container_name}
          rule {
            source_labels = ["__meta_kubernetes_namespace", "__meta_kubernetes_pod_container_name"]
            separator = "@"
            regex = "(.*)@(.*)"
            replacement = "ebpf/${1}/${2}"
            target_label = "service_name"
          }
        }

        pyroscope.ebpf "ebpf_pods" {
          targets = discovery.relabel.ebpf_pods.output
          demangle = "none"
          forward_to = argument.profiles_destinations.value
        }
      }
